### ISSUE TITLE
начала прове

### ISSUE THREAD

**Issue by withoutoxic** [1079374315/GT#5]
1) Быстрая установка на Linux (Ubuntu/Debian)
A) Системные пакеты
sudo apt update
sudo apt install -y python3 python3-venv python3-pip git ffmpeg libgl1

ffmpeg часто нужен для видео.

libgl1 спасает OpenCV от истерик “libGL.so.1 not found”.

B) Виртуальное окружение
python3 -m venv venv
source venv/bin/activate
pip install -U pip wheel
2) PyTorch на Linux: CPU или NVIDIA GPU
Вариант 1: CPU (самый простой, самый медленный)
pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu
Вариант 2: NVIDIA GPU (если есть CUDA)

Сначала проверь драйвер:

nvidia-smi

Если оно показывает твою видеокарту и версию драйвера, ставим PyTorch с CUDA (подбери команду под твою CUDA с официального сайта PyTorch, но обычно ок так):

pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121

Проверка в Python:

python -c "import torch; print(torch.cuda.is_available(), torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'CPU')"
3) Установка OpenCV и утилит
pip install opencv-python numpy
4) Структура проекта (по-человечески)

Создай папку и файлы:

mkdir -p video_ai/{data,src}
cd video_ai
touch src/{model.py,data.py,train.py,run.py}
5) Код под Linux (копируй как есть)
src/model.py
import torch
import torch.nn as nn
import torchvision.models as models

class VideoModel(nn.Module):
    def __init__(self, num_classes=10):
        super().__init__()

        resnet = models.resnet18(weights=models.ResNet18_Weights.DEFAULT)
        self.cnn = nn.Sequential(*list(resnet.children())[:-1])  # remove fc

        self.lstm = nn.LSTM(
            input_size=512,
            hidden_size=256,
            num_layers=2,
            batch_first=True
        )

        self.fc = nn.Linear(256, num_classes)

    def forward(self, x):
        # x: (B, T, C, H, W)
        b, t, c, h, w = x.shape
        x = x.view(b * t, c, h, w)
        feats = self.cnn(x)              # (B*T, 512, 1, 1)
        feats = feats.view(b, t, 512)    # (B, T, 512)

        out, _ = self.lstm(feats)        # (B, T, 256)
        logits = self.fc(out[:, -1])     # last timestep
        return logits
src/data.py
import cv2
import numpy as np
import torch
from torch.utils.data import Dataset

def load_video_frames(path, num_frames=16, size=224):
    cap = cv2.VideoCapture(path)
    if not cap.isOpened():
        raise RuntimeError(f"Cannot open video: {path}")

    total = int(cap.get(cv2.CAP_PROP_FRAME_COUNT)) or 0
    step = max(total // num_frames, 1) if total > 0 else 1

    frames = []
    for i in range(num_frames):
        cap.set(cv2.CAP_PROP_POS_FRAMES, i * step)
        ret, frame = cap.read()
        if not ret:
            break
        frame = cv2.resize(frame, (size, size))
        frame = frame[:, :, ::-1]  # BGR->RGB
        frame = frame.astype(np.float32) / 255.0
        frame = np.transpose(frame, (2, 0, 1))  # C,H,W
        frames.append(frame)

    cap.release()

    if len(frames) == 0:
        raise RuntimeError(f"No frames read from: {path}")

    # pad if short
    while len(frames) < num_frames:
        frames.append(frames[-1])

    arr = np.stack(frames, axis=0)  # (T,C,H,W)
    return torch.tensor(arr, dtype=torch.float32)

class SimpleVideoDataset(Dataset):
    """
    Ожидает список (path, label).
    """
    def __init__(self, items, num_frames=16):
        self.items = items
        self.num_frames = num_frames

    def __len__(self):
        return len(self.items)

    def __getitem__(self, idx):
        path, label = self.items[idx]
        x = load_video_frames(path, num_frames=self.num_frames)
        y = torch.tensor(label, dtype=torch.long)
        return x, y
src/train.py
import torch
from torch.utils.data import DataLoader
import torch.nn as nn
import torch.optim as optim

from src.model import VideoModel
from src.data import SimpleVideoDataset

def train(train_items, num_classes, epochs=3, batch_size=2, lr=1e-4, device=None):
    device = device or ("cuda" if torch.cuda.is_available() else "cpu")
    print("Device:", device)

    ds = SimpleVideoDataset(train_items, num_frames=16)
    dl = DataLoader(ds, batch_size=batch_size, shuffle=True, num_workers=2, pin_memory=True)

    model = VideoModel(num_classes=num_classes).to(device)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=lr)

    model.train()
    for ep in range(1, epochs + 1):
        total_loss = 0.0
        correct = 0
        seen = 0

        for x, y in dl:
            x = x.to(device)         # (B,T,C,H,W)
            y = y.to(device)

            optimizer.zero_grad()
            logits = model(x)
            loss = criterion(logits, y)
            loss.backward()
            optimizer.step()

            total_loss += loss.item() * x.size(0)
            pred = logits.argmax(dim=1)
            correct += (pred == y).sum().item()
            seen += x.size(0)

        print(f"Epoch {ep}: loss={total_loss/seen:.4f} acc={correct/seen:.3f}")

    return model
src/run.py (инференс)
import torch
from src.model import VideoModel
from src.data import load_video_frames

def predict(video_path, weights_path, num_classes):
    device = "cuda" if torch.cuda.is_available() else "cpu"

    model = VideoModel(num_classes=num_classes).to(device)
    model.load_state_dict(torch.load(weights_path, map_location=device))
    model.eval()

    x = load_video_frames(video_path, num_frames=16).unsqueeze(0).to(device)
    with torch.no_grad():
        logits = model(x)
        cls = logits.argmax(dim=1).item()
    return cls

if __name__ == "__main__":
    print(predict("video.mp4", "weights.pt", num_classes=5))
6) Запуск обучения (пример)

Создай маленький список данных прямо в Python (потом заменишь на реальный датасет):

python - <<'PY'
from src.train import train
# Пример: 2 видео, 2 класса (0/1). Поменяй пути на реальные mp4.
items = [
    ("data/a.mp4", 0),
    ("data/b.mp4", 1),
]
model = train(items, num_classes=2, epochs=2, batch_size=1)

import torch
torch.save(model.state_dict(), "weights.pt")
print("Saved weights.pt")
PY
7) Типичные Linux-грабли (чтобы ты не страдал лишний раз)

OpenCV ругается на libGL → sudo apt install libgl1

Видео не читается → проверь ffmpeg -i file.mp4

CUDA есть, но torch не видит → чаще всего несовпадение версии драйвера/torch-cuda, или установлен CPU torch

Если тебе реально нужно генерировать видео нейросетью (а не классифицировать), то это уже другой зверь: diffusion/transformer, VRAM нужен как воздух. Но под Linux тоже делается.
